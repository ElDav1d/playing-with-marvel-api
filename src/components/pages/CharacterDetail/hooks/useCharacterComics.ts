import { MAX_CHARACTER_COMICS } from '@/utils/constants';
import { useQuery } from '@tanstack/react-query';
import { getCharacterComicsService } from '../services';
MAX_CHARACTER_COMICS;

const useCharacterComics = (characterId: string | undefined, maxComics: number, page: number) => {
  const { isError, data, isFetching, refetch } = useQuery(
    ['characterComics', characterId],
    () => getCharacterComicsService({ characterId, maxComics, page }),
    {
      refetchOnWindowFocus: false,
      keepPreviousData: true,
    },
  );

  const safeOffset = data ? data.offset : 0;
  const comics = data?.response?.results;

  return {
    comics,
    totalComics: data?.response.total,
    rangeInit: safeOffset + 1,
    rangeEnd: safeOffset + comics?.length,
    isErrorOnComics: isError,
    isFetchingComics: isFetching,
    refetch,
  };
};

export default useCharacterComics;
