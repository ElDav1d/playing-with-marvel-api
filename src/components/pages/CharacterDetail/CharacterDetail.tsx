import { Link, useParams } from 'react-router-dom';
import { useCharacterDetails, useCharacterComics } from './hooks';
import Character from '@/components/organisms/Character/Character';
import { ComicsList } from '@/components/organisms/ComicsList/ComicsList';
import { MAX_CHARACTER_COMICS } from '@/utils/constants';
import { useState } from 'react';

const CharacterDetail = () => {
  const { id } = useParams();
  const [page, setPage] = useState<number>(0);

  const { isLoadingCharacter, isErrorOnCharacter, character } = useCharacterDetails(id);

  const { isErrorOnComics, comics, isFetchingComics, refetch } = useCharacterComics(
    id,
    MAX_CHARACTER_COMICS,
    page,
  );

  const handlePrevPage = () => {
    setPage(page - 1);
    refetch();
  };

  const handleNextPage = () => {
    setPage(page + 1);
    refetch();
  };

  const isError = isErrorOnCharacter || isErrorOnComics;

  return (
    <>
      {isLoadingCharacter && <h2>Loading Character Details...</h2>}
      {character && (
        <Character
          name={character.name}
          description={character.description}
          thumbnailPath={character.thumbnail.path}
          thumbnailExtension={character.thumbnail.extension}
        />
      )}
      {isFetchingComics && <h2>Loading Character Comics</h2>}
      {comics && <ComicsList comics={comics} />}
      {isError && <h2>Ooops, try reloading again</h2>}
      <button onClick={handlePrevPage}>Previous Comics</button>
      <button onClick={handleNextPage}>Next Comics</button>
      <Link to='/'>Home</Link>
    </>
  );
};

export default CharacterDetail;
