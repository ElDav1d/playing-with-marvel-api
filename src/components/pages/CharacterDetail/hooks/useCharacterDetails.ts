import { useQuery } from '@tanstack/react-query';
import { getCharacterDetailsService } from '../services';
import CharacterDetail from '../CharacterDetail';

/**
 * Props for the useCharacterDetails custom hook.
 * @interface
 */
export interface IUseCharacterDetailsProps {
  /**
   * The character ID.
   * @type {string}
   * @default undefined
   */
  characterId: string | undefined;
}
/**
 * Return type for the useCharacterDetails custom hook.
 * @interface
 */
export interface IUseCharacterDetailsReturn {
  /**
   * Indicates if the character is loading.
   * @property {boolean}
   */
  isLoadingCharacter: boolean;
  /**
   * Indicates if there is an error on the character.
   * @property {boolean}
   */
  isErrorOnCharacter: boolean;
  /**
   * The character details.
   * @property {CharacterDetails | undefined}
   */
  character: typeof CharacterDetail | undefined;
}

/**
 * Custom hook to fetch character details.
 * @function
 * @param {IUseCharacterDetailsProps} props - The props for the hook.
 * @returns {IUseCharacterDetailsReturn}
 */
const useCharacterDetails = ({
  characterId,
}: IUseCharacterDetailsProps): IUseCharacterDetailsReturn => {
  const { isLoading, isError, data } = useQuery({
    queryKey: ['characterDetails', characterId],
    queryFn: () => getCharacterDetailsService(characterId),
    refetchOnWindowFocus: false,
  });

  return {
    isLoadingCharacter: isLoading,
    isErrorOnCharacter: isError,
    character: data,
  };
};

export default useCharacterDetails;
