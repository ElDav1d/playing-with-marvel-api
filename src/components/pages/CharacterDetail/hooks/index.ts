import useCharacterComics from './useCharacterComics';
import useCharacterDetails from './useCharacterDetails';

export { useCharacterDetails, useCharacterComics };
