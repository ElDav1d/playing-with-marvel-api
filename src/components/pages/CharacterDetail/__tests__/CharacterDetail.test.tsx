import mockCharacterDetail from '../mocks/mockCharacterDetail.json';
import mockCharacterComics from '../mocks/mockCharacterComics.json';
import { RenderOptions, render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import CharacterDetail from '../CharacterDetail';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { useCharacterDetails, useCharacterComics } from '../hooks';
import userEvent from '@testing-library/user-event';

jest.mock('../hooks');

const queryClient = new QueryClient();

jest.mock('react-lazy-load-image-component', () => ({
  LazyLoadImage: () => null,
}));

jest.mock('../../../molecules/SelectGroup', () => {
  return {
    EmotionCacheProvider: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  };
});

const mockUseCharacterDetails = useCharacterDetails as jest.Mock;

const mockUseCharacterComics = useCharacterComics as jest.Mock;

const renderSnapshot = (ui: React.ReactElement, options?: RenderOptions) => {
  return render(<QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>, options);
};

it('renders the character detail page and matches snapshot', () => {
  // ARRANGE
  mockUseCharacterDetails.mockReturnValue({
    isLoadingCharacter: false,
    isErrorOnCharacter: false,
    character: mockCharacterDetail,
  });

  mockUseCharacterComics.mockReturnValue({
    comics: mockCharacterComics,
    totalComics: 0,
    rangeInit: 0,
    rangeEnd: 0,
    isErrorOnComics: false,
    isFetchingComics: false,
    isFirstPage: true,
    isLastPage: true,
    refetch: jest.fn(),
  });

  // ACT
  const { asFragment } = renderSnapshot(
    <QueryClientProvider client={queryClient}>
      <Router>
        <CharacterDetail />
      </Router>
    </QueryClientProvider>,
  );

  // ASSERT
  expect(asFragment()).toMatchSnapshot();
});

it('renders the loader when fetching initial page', () => {
  // ARRANGE
  mockUseCharacterDetails.mockReturnValue({
    isLoadingCharacter: true,
    isErrorOnCharacter: false,
    character: undefined,
  });

  mockUseCharacterComics.mockReturnValue({
    comics: [],
    totalComics: 0,
    rangeInit: 0,
    rangeEnd: 0,
    isErrorOnComics: false,
    isFetchingComics: false,
    isFirstPage: true,
    isLastPage: true,
    refetch: jest.fn(),
  });

  // ACT
  render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <CharacterDetail />
      </Router>
    </QueryClientProvider>,
  );

  // ASSERT
  expect(screen.getByRole('alert', { name: /(loading)/i })).toBeInTheDocument();
});

it('renders the character detail page with appropiate elements', () => {
  // ARRANGE
  mockUseCharacterDetails.mockReturnValue({
    isLoadingCharacter: false,
    isErrorOnCharacter: false,
    character: mockCharacterDetail,
  });

  mockUseCharacterComics.mockReturnValue({
    comics: mockCharacterComics,
    totalComics: 0,
    rangeInit: 0,
    rangeEnd: 0,
    isErrorOnComics: false,
    isFetchingComics: false,
    isFirstPage: true,
    isLastPage: true,
    refetch: jest.fn(),
  });

  // ACT
  render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <CharacterDetail />
      </Router>
    </QueryClientProvider>,
  );

  // ASSERT
  expect(screen.getByRole('banner', { name: /common header/i })).toBeInTheDocument();
  expect(screen.getByRole('navigation', { name: /main navigation/i })).toBeInTheDocument();
  expect(screen.getByRole('link', { name: /playing with marvel api/i })).toBeInTheDocument();
  expect(screen.getByRole('main', { name: /character detail main content/i })).toBeInTheDocument();
  expect(screen.getByRole('article', { name: /character detail article/i })).toBeInTheDocument();
  expect(screen.getByRole('contentinfo', { name: /common footer/i })).toBeInTheDocument();
  expect(screen.getByRole('heading', { level: 2, name: /abomination/i })).toBeInTheDocument();
});

it('renders a list of comics when the character has some', () => {
  // ARRANGE
  mockUseCharacterDetails.mockReturnValue({
    isLoadingCharacter: false,
    isErrorOnCharacter: false,
    character: mockCharacterDetail,
  });

  mockUseCharacterComics.mockReturnValue({
    comics: mockCharacterComics,
    totalComics: 0,
    rangeInit: 0,
    rangeEnd: 0,
    isErrorOnComics: false,
    isFetchingComics: false,
    isFirstPage: true,
    isLastPage: true,
    refetch: jest.fn(),
  });

  // ACT
  render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <CharacterDetail />
      </Router>
    </QueryClientProvider>,
  );

  // ASSERT
  expect(screen.getByRole('list', { name: /list of comics/i })).toBeInTheDocument();
  expect(screen.getAllByRole('listitem')).toHaveLength(10);
});

it('renders an order selector for the comics list when the character has some', () => {
  // ARRANGE
  mockUseCharacterDetails.mockReturnValue({
    isLoadingCharacter: false,
    isErrorOnCharacter: false,
    character: mockCharacterDetail,
  });

  mockUseCharacterComics.mockReturnValue({
    comics: mockCharacterComics,
    totalComics: 0,
    rangeInit: 0,
    rangeEnd: 0,
    isErrorOnComics: false,
    isFetchingComics: false,
    isFirstPage: true,
    isLastPage: true,
    refetch: jest.fn(),
  });

  // ACT
  render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <CharacterDetail />
      </Router>
    </QueryClientProvider>,
  );

  // ASSERT
  expect(screen.getByRole('combobox', { name: /order comics/i })).toBeInTheDocument();
});

it('render the comics list order selector with the required options', () => {
  // ARRANGE
  mockUseCharacterDetails.mockReturnValue({
    isLoadingCharacter: false,
    isErrorOnCharacter: false,
    character: mockCharacterDetail,
  });

  mockUseCharacterComics.mockReturnValue({
    comics: mockCharacterComics,
    totalComics: 0,
    rangeInit: 0,
    rangeEnd: 0,
    isErrorOnComics: false,
    isFetchingComics: false,
    isFirstPage: true,
    isLastPage: true,
    refetch: jest.fn(),
  });

  // ACT
  render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <CharacterDetail />
      </Router>
    </QueryClientProvider>,
  );

  const selectOptionTitleAZ = screen.getByRole('option', {
    name: /title \(a\/z\)/i,
  }) as HTMLOptionElement;
  const selectOptionTitleZA = screen.getByRole('option', {
    name: /title \(z\/a\)/i,
  }) as HTMLOptionElement;
  const selectOptionFirstLast = screen.getByRole('option', {
    name: /(first\/last)+/i,
  }) as HTMLOptionElement;
  const selectOptionLastFirst = screen.getByRole('option', {
    name: /(last\/first)+/i,
  }) as HTMLOptionElement;

  // ASSERT
  expect(selectOptionTitleAZ.value).toBe('title');
  expect(selectOptionTitleZA.value).toBe('-title');
  expect(selectOptionFirstLast.value).toBe('issueNumber');
  expect(selectOptionLastFirst.value).toBe('-issueNumber');
});

it('fetches a comic list when order is changed', async () => {
  // ARRANGE
  const user = userEvent.setup();

  mockUseCharacterDetails.mockReturnValue({
    isLoadingCharacter: false,
    isErrorOnCharacter: false,
    character: mockCharacterDetail,
  });

  mockUseCharacterComics.mockReturnValue({
    comics: mockCharacterComics,
    totalComics: 0,
    rangeInit: 0,
    rangeEnd: 0,
    isErrorOnComics: false,
    isFetchingComics: false,
    isFirstPage: true,
    isLastPage: true,
    refetch: jest.fn(),
  });

  // ACT
  render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <CharacterDetail />
      </Router>
    </QueryClientProvider>,
  );

  // reordering not happening because items are mocked
  user.selectOptions(screen.getByRole('combobox', { name: /order comics/i }), '-title');

  // ASSERT
  await waitFor(() => {
    expect(mockUseCharacterComics).toHaveBeenCalledTimes(3);
    expect(screen.getByRole('list', { name: /list of comics/i })).toBeInTheDocument();
  });
});
