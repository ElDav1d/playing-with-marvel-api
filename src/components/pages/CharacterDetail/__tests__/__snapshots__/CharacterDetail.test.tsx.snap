// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the character detail page and matches snapshot 1`] = `
<DocumentFragment>
  <header
    aria-label="common header"
    class="w-full fixed z-1 top-0 flex justify-center transition-[transform] delay-short duration-header  bg-black "
  >
    <nav
      aria-label="main navigation"
      class="mx-6 md:mx-12 lg:mx-18 xl:mx-auto max-w-[1240px]"
    >
      <a
        class="block focus-visible"
        href="/"
      >
        <h1
          class="text-none"
        >
          Playing with Marvel API
          <svg
            aria-hidden="true"
            height="52"
            role="img"
            width="130"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              MARVEL logo
            </title>
            <rect
              fill="#EC1D24"
              height="100%"
              width="100%"
            />
            <path
              d="M126.222 40.059v7.906H111.58V4h7.885v36.059h6.757zm-62.564-14.5c-.61.294-1.248.44-1.87.442v-14.14h.04c.622-.005 5.264.184 5.264 6.993 0 3.559-1.58 5.804-3.434 6.705zM40.55 34.24l2.183-18.799 2.265 18.799H40.55zm69.655-22.215V4.007H87.879l-3.675 26.779-3.63-26.78h-8.052l.901 7.15c-.928-1.832-4.224-7.15-11.48-7.15-.047-.002-8.06 0-8.06 0l-.031 39.032-5.868-39.031-10.545-.005-6.072 40.44.002-40.435H21.278L17.64 26.724 14.096 4.006H4v43.966h7.95V26.78l3.618 21.192h4.226l3.565-21.192v21.192h15.327l.928-6.762h6.17l.927 6.762 15.047.008h.01v-.008h.02V33.702l1.845-.27 3.817 14.55h7.784l-.002-.01h.022l-5.011-17.048c2.538-1.88 5.406-6.644 4.643-11.203v-.002C74.894 19.777 79.615 48 79.615 48l9.256-.027 6.327-39.85v39.85h15.007v-7.908h-7.124v-10.08h7.124v-8.03h-7.124v-9.931h7.124z"
              fill="#FEFEFE"
            />
            <path
              d="M0 0h30v52H0z"
              fill="#EC1D24"
            />
            <path
              d="M31.5 48V4H21.291l-3.64 22.735L14.102 4H4v44h8V26.792L15.577 48h4.229l3.568-21.208V48z"
              fill="#FEFEFE"
            />
          </svg>
        </h1>
      </a>
    </nav>
  </header>
  <main
    aria-label="character detail main content"
    class="pb-48 min-h-[100vh]"
  >
    <article
      aria-label="character detail article"
    >
      <section
        class="relative lg:h-[65vh] overflow-hidden pt-logoDefaultHeight bg-stone-900 text-white clip-hero md:before:absolute md:before:inset-0 md:before:content[\\"\\"] md:before:bg-gradient-to-r md:before:from-trans-0.75-black md:before:to-transparent md:before:z-1"
      >
        <picture
          class="md:block: md:h-full flex items-center justify-center"
        />
        <div
          class="w-full h-full flex flex-col justify-center py-14 md:py-0 md:absolute md:top-0 md:left-0 md:z-10"
        >
          <div
            class="mx-6 md:mx-12 lg:mx-18 xl:mx-auto max-w-[1240px] w-full"
          >
            <h2
              class="text-2xl uppercase font-semibold mb-4"
            >
              Abomination (Emil Blonsky)
            </h2>
            <p>
              Formerly known as Emil Blonsky, a spy of Soviet Yugoslavian origin working for the KGB, the Abomination gained his powers after receiving a dose of gamma radiation similar to that which transformed Bruce Banner into the incredible Hulk.
            </p>
          </div>
        </div>
      </section>
      <section
        class="mx-6 md:mx-12 lg:mx-18 xl:mx-auto max-w-[1240px]"
      >
        <fieldset
          class=""
        >
          <legend
            class="mb-1"
          >
            Order comics by:
          </legend>
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class=" css-13cymwt-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-2-placeholder"
                >
                  Select...
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-2-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </fieldset>
        <h3
          class="mb-2"
        >
          Displaying 0 to 0 from 0 available comics
        </h3>
        <ul
          aria-label="List of comics for the character."
          aria-live="polite"
          class="grid gap-3 grid-flow-row grid-cols-auto-min-max-120-auto md:grid-cols-auto-min-max-185-auto"
        >
          <li
            aria-label="Maestro: World War M (2022) #5"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                Maestro: World War M (2022) #5
              </h3>
              <p
                class="text-xs leading-5 line-clamp-3"
              >
                When you can live forever, what does it mean to live? In this shocking series finale, Maestro finds himself partnered with an old enemy against even older allies to determine once and for all the true ruler of Dystopia. But opportunism depends on knowing how your opponents think, and someone's failed to account for a few tricks…
              </p>
            </div>
          </li>
          <li
            aria-label="Maestro: World War M (2022) #3"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                Maestro: World War M (2022) #3
              </h3>
              <p
                class="text-xs leading-5 line-clamp-3"
              >
                Maestro clashes with the Abomination in a showdown for the ages! But Abomination is far from the only person gunning for Maestro, and when Emil Blonsky turns out to be just the opening salvo of a larger attack, can the Maestro continue to out-smash and outwit the forces that want him dead?
              </p>
            </div>
          </li>
          <li
            aria-label="MARVEL MASTERWORKS: THE INCREDIBLE HULK VOL. 11 HC (Trade Paperback)"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                MARVEL MASTERWORKS: THE INCREDIBLE HULK VOL. 11 HC (Trade Paperback)
              </h3>
              <p
                class="text-xs leading-5 line-clamp-3"
              >
                Collecting Incredible Hulk (1968) #184-196 and material from Giant-Size Hulk (1975) #1 and Marvel Treasury Edition #5. Hulk is the strongest! Why? Because it's hard not to be when you go from strength to strength with artists Herb Trimpe and Sal Buscema! Trimpe defined the incredible Hulk in an artistic tenure stretching from 1968 to 1975. When Sal Buscema took over the reins, not only did the series gain a great talent, but one that, unbelievably, would have a tenure longer than even Trimpe's! Add to that Len Wein writing some of his greatest Hulk stories and you've got a bona fi de Marvel Masterworks! The adventures include Hulk smashing his way through the Mole Man, the Gremlin, the Shaper of Worlds, Doc Samson, the Abomination...and, of course, the never-ending military machinations of Gen. "Thunderbolt" Ross!
              </p>
            </div>
          </li>
          <li
            aria-label="Incredible Hulk Epic Collection: The Hulk Must Die (Trade Paperback)"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                Incredible Hulk Epic Collection: The Hulk Must Die (Trade Paperback)
              </h3>
              <p
                class="text-xs leading-5 line-clamp-3"
              >
                Marvel’s Green Goliath is pulling no punches! After the cancellation of his debut series, the Incredible Hulk returns in TALES TO ASTONISH — and he means business! Dr. Robert Bruce Banner’s alter ego battles General “Thunderbolt” Ross, the Leader and his Humanoids, the Executioner, and more Communists than you can shake a tank at. And if that’s not enough, round two pits the Hulk against Hercules, introduces classic villain the Abomination, dives into the intrigue of the Secret Empire, and goes cosmic with the Silver Surfer and the High Evolutionary! Featuring stories by Stan Lee and art by a who’s who of Marvel greats — Steve Ditko, Jack Kirby, Gil Kane, Bill Everett and John Buscema — these are the tales that made the Hulk a global icon! Collecting material from TALES TO ASTONISH (1959) #60-96 and NOT BRAND ECHH #3.
              </p>
            </div>
          </li>
          <li
            aria-label="Marvel Cinematic Universe Guidebook: The Avengers Initiative (Trade Paperback)"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                Marvel Cinematic Universe Guidebook: The Avengers Initiative (Trade Paperback)
              </h3>
              <p
                class="text-xs leading-5 line-clamp-3"
              >
                The official handbook of the Marvel Cinematic Universe takes you all the way through Phase One and to the dawn of Phase Two! Filled with fact sheets, movie-to-comic comparisons, behind-the-scenes art and production stills, this volume details the first eight blockbuster Marvel movies — from 2008’s Iron Man to 2013’s Thor: The Dark World. In-depth character profiles give you the whole story on heroes like Captain America, Hulk, Black Widow, War Machine, Hawkeye and more; deadly villains including Abomination, Loki, Red Skull and Mandarin; and a universe of supporting characters such as Nick Fury, Bucky Barnes, Peggy Carter and Phil Coulson! It’s your indispensable MCU guide! Collecting GUIDEBOOK TO THE MARVEL CINEMATIC UNIVERSE — MARVEL’S IRON MAN, MARVEL’S INCREDIBLE HULK, MARVEL’S IRON MAN 2, MARVEL’S THOR, MARVEL’S CAPTAIN AMERICA: THE FIRST AVENGER, MARVEL’S THE AVENGERS, MARVEL’S IRON MAN 3 and MARVEL’S THOR: THE DARK WORLD.
              </p>
            </div>
          </li>
          <li
            aria-label="Incredible Hulks (2010) #619"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                Incredible Hulks (2010) #619
              </h3>
            </div>
          </li>
          <li
            aria-label="Incredible Hulks (2010) #618"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                Incredible Hulks (2010) #618
              </h3>
            </div>
          </li>
          <li
            aria-label="Incredible Hulks (2010) #602 (SHS VARIANT)"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                Incredible Hulks (2010) #602 (SHS VARIANT)
              </h3>
            </div>
          </li>
          <li
            aria-label="Hulk (2008) #3"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                Hulk (2008) #3
              </h3>
            </div>
          </li>
          <li
            aria-label="IRREDEEMABLE ANT-MAN VOL. 1: LOW-LIFE DIGEST (Trade Paperback)"
            class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
          >
            <picture
              class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
            />
            <div
              class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-short before:duration-gridItem before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
            >
              <h3
                class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
              >
                IRREDEEMABLE ANT-MAN VOL. 1: LOW-LIFE DIGEST (Trade Paperback)
              </h3>
            </div>
          </li>
        </ul>
      </section>
    </article>
  </main>
  <footer
    aria-label="common footer"
    class="absolute bottom-0 w-full p-6 bg-stone-900"
  >
    <section
      class="mx-6 md:mx-12 lg:mx-18 xl:mx-auto max-w-[1240px]"
    >
      <a
        class="block active focus-visible"
        href="/"
      >
        <span
          aria-hidden="true"
        >
          <svg
            aria-hidden="true"
            height="115"
            viewBox="0 0 36 52"
            width="80"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              MARVEL only M logo
            </title>
            <rect
              fill="#EC1D24"
              height="100%"
              width="100%"
            />
            <path
              d="M31.5 48V4H21.291l-3.64 22.735L14.102 4H4v44h8V26.792L15.577 48h4.229l3.568-21.208V48z"
              fill="#FEFEFE"
            />
          </svg>
        </span>
      </a>
    </section>
  </footer>
</DocumentFragment>
`;
